File: DSA/Navneetpant_590011569/Day_72/Q2.java
import java.util.*;

class Solution {

    static class Edge {
        int u, v, w;

        Edge(int u, int v, int w) {
            this.u = u;
            this.v = v;
            this.w = w;
        }
    }

    static int[] parent;
    static int[] rank;

    static int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    static boolean union(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        if (rootA == rootB) {
            return false;
        }

        if (rank[rootA] < rank[rootB]) {
            parent[rootA] = rootB;
        } else if (rank[rootA] > rank[rootB]) {
            parent[rootB] = rootA;
        } else {
            parent[rootB] = rootA;
            rank[rootA]++;
        }

        return true;
    }

    public static int kruskals(int gNodes, List<Integer> gFrom,
                               List<Integer> gTo, List<Integer> gWeight) {

        int m = gFrom.size();

        List<Edge> edges = new ArrayList<>();

        for (int i = 0; i < m; i++) {
            edges.add(new Edge(
                gFrom.get(i),
                gTo.get(i),
                gWeight.get(i)
            ));
        }

        // Sort by weight, then by u + v + w
        edges.sort((a, b) -> {
            if (a.w != b.w) {
                return Integer.compare(a.w, b.w);
            }

            int sumA = a.u + a.v + a.w;
            int sumB = b.u + b.v + b.w;

            return Integer.compare(sumA, sumB);
        });

        parent = new int[gNodes + 1];
        rank = new int[gNodes + 1];

        for (int i = 1; i <= gNodes; i++) {
            parent[i] = i;
        }

        int totalWeight = 0;
        int edgesUsed = 0;

        for (Edge edge : edges) {

            if (union(edge.u, edge.v)) {
                totalWeight += edge.w;
                edgesUsed++;

                if (edgesUsed == gNodes - 1) {
                    break;
                }
            }
        }

        return totalWeight;
    }
}

File: DSA/VivaswanSingh_590023022/Day_72/q2.c
#include <stdio.h>
#include <stdlib.h>

typedef struct
{
    int u;
    int v;
    int weight;
} Edge;

int parent[1000];

int compare(const void *a, const void *b)
{
    Edge *x = (Edge *)a;
    Edge *y = (Edge *)b;

    return x->weight - y->weight;
}

int find(int x)
{
    if (parent[x] != x)
        parent[x] = find(parent[x]);

    return parent[x];
}

int unite(int a, int b)
{
    int rootA = find(a);
    int rootB = find(b);

    if (rootA == rootB)
        return 0;

    parent[rootB] = rootA;
    return 1;
}

int kruskal(int n, int m, Edge edges[])
{
    for (int i = 1; i <= n; i++)
        parent[i] = i;

    qsort(edges, m, sizeof(Edge), compare);

    int totalWeight = 0;
    int edgesUsed = 0;

    for (int i = 0; i < m && edgesUsed < n - 1; i++)
    {
        if (unite(edges[i].u, edges[i].v))
        {
            totalWeight += edges[i].weight;
            edgesUsed++;
        }
    }

    return totalWeight;
}

int main()
{
    int n = 4;
    int m = 6;

    Edge edges[] = {
        {1, 2, 5},
        {1, 3, 3},
        {4, 1, 6},
        {2, 4, 7},
        {3, 2, 4},
        {3, 4, 5}
    };

    printf("%d\n", kruskal(n, m, edges));

    return 0;
}